src/application/controller/controller.c

```diff
@@ -1,2 +1,166 @@
+    // Build the CAN msg using [canard-specific canlib function to be defined
+    // later]. Send this to can handler module’s tx
+    return W_SUCCESS;
+}
+
+/**
+ * Initialize controller module
+ * Must be called before RTOS scheduler starts
+ * @return W_SUCCESS if initialization successful
+ */
+w_status_t controller_init(void) {
+    w_status_t init_status = W_SUCCESS;
+
+    // Create internal state queue (length = 1)
+    internal_state_queue =
+        xQueueCreate(1, sizeof(controller_input_t)); // estimator function to controller task
+    output_queue = xQueueCreate(1, sizeof(controller_output_t));
+
+    // check queue creation
+    if (internal_state_queue == NULL || output_queue == NULL) {
+        init_status = W_FAILURE;
+        log_text("controller", "queue creation failed");
+    }
+
+    // gain table file memory test
+    gain_table = fopen("file path", "r");
+
+    if (gain_table == NULL) {
+        log_text("controller", "gain table invalid");
+        init_status = W_FAILURE;
+    } else {
+        log_text("controller", "gain table file opened");
+    }
+
+    // Log initialization status
+    if (init_status == W_SUCCESS) {
+        log_text("controller", "initialization successful");
+
+    } else {
+        log_text("controller", "initialization failed");
+    }
+
+    // return w_status_t state
+    return init_status;
+}
+
+/**
+ * Update controller with new state data - called by state estimation module
+ * @param new_state Latest state estimate from state estimation
+ * @return W_FAILURE if validation/queueing fails
+ */
+w_status_t controller_update_inputs(controller_input_t *new_state) {
+    xQueueOverwrite(internal_state_queue, new_state); // overwrite internal queue
+    log_text("controller", "latest output received, timestamp %d", new_state->timestamp);
+    return W_SUCCESS;
+}
+
+/**
+ * Get most recent control output (cammanded angle and timestamp) - called by
+ * state estimation module
+ * @param output Pointer to store output -> type defined in controller
+ * @return W_FAILURE if no output available
+ */
+w_status_t controller_get_latest_output(controller_output_t *output) {
+    if (xQueuePeek(output_queue, output, timeout) == pdPASS) {
+        log_text("controller", "latest output returned");
+        return W_SUCCESS;
+    }
+    log_text("controller", "timeout: no output available");
+    return W_FAILURE;
+}
+
+/**
+ * Controller task function for RTOS
+ */
+void controller_task(void *argument) {
+    // get current flight phase
+    flight_phase_state_t current_phase = STATE_INIT;
+
+    while (true) {
+        // log phase transitions, specifics logged in flight phase
+        if (current_phase != flight_phase_get_state()) {
+            current_phase = flight_phase_get_state();
+            log_text("controller", "flight phase changed");
+        }
+
+        if (current_phase != STATE_ACT_ALLOWED &&
+            current_phase != STATE_COAST) { // if not in proper state
```

Comment:
`STATE_COAST` is also fake. `STATE_ACT_ALLOWED ` is the only valid state for controller actuation. That one is on me oops